src/shims/env.rs

```diff
@@ -27,24 +28,24 @@ impl EnvVars {
             for (name, value) in env::vars() {
                 if !excluded_env_vars.contains(&name) {
                     let var_ptr =
-                        alloc_env_var(name.as_bytes(), value.as_bytes(), &mut ecx.memory);
-                    ecx.machine.env_vars.map.insert(name.into_bytes(), var_ptr);
+                        alloc_env_var_as_c_str(name.as_bytes(), value.as_bytes(), ecx);
+                    ecx.machine.env_vars.map.insert(OsString::from(name), var_ptr);
                 }
             }
         }
     }
 }
 
-fn alloc_env_var<'mir, 'tcx>(
+fn alloc_env_var_as_c_str<'mir, 'tcx>(
     name: &[u8],
     value: &[u8],
-    memory: &mut Memory<'mir, 'tcx, Evaluator<'tcx>>,
+    ecx: &mut InterpCx<'mir, 'tcx, Evaluator<'tcx>>
 ) -> Pointer<Tag> {
     let mut bytes = name.to_vec();
     bytes.push(b'=');
     bytes.extend_from_slice(value);
     bytes.push(0);
-    memory.allocate_static_bytes(bytes.as_slice(), MiriMemoryKind::Env.into())
+    ecx.alloc_os_str_as_c_str(bytes_to_os_str(bytes.as_slice()).unwrap())
```

Comment:
Instead of calling `bytes_to_os_str` and its inverse everywhere, try converting as much of this file as you can to `OsStr` instead of `&[u8]`. This file here should basically not work with byte slices at all any more.